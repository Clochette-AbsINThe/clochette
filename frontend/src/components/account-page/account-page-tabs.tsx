import Link from 'next/link';
import { useRouter } from 'next/router';
import type { Session } from 'next-auth';
import { useSession } from 'next-auth/react';

import { SecurityScopes } from '@/openapi-codegen/clochetteSchemas';
import { pages } from '@/utils/pages';

type LinksType = Array<{
  label: string;
  href: string;
  scopes: SecurityScopes[];
}>;

export const links: LinksType = [
  { label: 'Accueil', href: pages.account.index, scopes: ['staff', 'treasurer', 'president'] },
  { label: 'Tableau de bord', href: pages.account.dashboard, scopes: ['president', 'treasurer'] },
  { label: 'Comptes utilisateurs', href: pages.account.users, scopes: ['president'] },
  { label: 'TrÃ©sorie', href: pages.account.tresory, scopes: ['president', 'treasurer'] }
];

export function findActiveLink(pathname: string) {
  return links.find((link) => link.href.search(pathname) !== -1);
}

export function filterLinks(session: Session | null) {
  if (session) {
    const userRoles = session.scopes;
    return links.filter((link) => link.scopes.some((role) => userRoles.includes(role)));
  } else {
    return [];
  }
}

export function Tabs() {
  const { pathname } = useRouter();
  const { data: session } = useSession();

  const activeLink = findActiveLink(pathname);
  const filteredLinks = filterLinks(session);

  return (
    <div className='flex justify-center pr-2 md:border-r md:border-b-0 border-b md:mr-4 mb-4 md:mb-0'>
      <div className='w-full max-w-md md:w-max flex justify-start items-center md:items-start md:flex-col flex-row mb-2 md:mb-0'>
        {filteredLinks.map((link) => (
          <Link
            href={link.href}
            key={link.href}
            className={`px-3 py-2 text-sm w-full font-medium ${activeLink === link ? 'bg-slate-400 dark:bg-gray-800' : 'text-gray-900 hover:bg-slate-300 dark:text-gray-300 dark:hover:bg-gray-700'}`}
          >
            {link.label}
          </Link>
        ))}
      </div>
    </div>
  );
}
