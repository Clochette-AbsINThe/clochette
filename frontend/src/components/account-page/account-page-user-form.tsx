import { useForm } from 'react-hook-form';
import toast from 'react-hot-toast';

import { useQueryClient } from '@tanstack/react-query';

import { Button } from '@/components/button';
import { Form, FormControl, FormDescription, FormField, FormItem, FormLabel, FormMessage } from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { useReadAccountMe, useUpdateAccountMe } from '@/openapi-codegen/clochetteComponents';
import { generateApiErrorMessage } from '@/openapi-codegen/clochetteFetcher';
import { Account, OwnAccountUpdate } from '@/openapi-codegen/clochetteSchemas';

interface UserFormProps {
  user: Account;
}

export function UserForm(props: UserFormProps): React.JSX.Element {
  const query = useQueryClient();

  const { data: user } = useReadAccountMe(
    {},
    {
      initialData: props.user
    }
  );

  const updateOwnAccount = useUpdateAccountMe({
    onSuccess: () => {
      toast.success('Votre compte a été mis à jour.');
      query.invalidateQueries({ stale: true });
    },
    onError: (error) => {
      const detail = generateApiErrorMessage(error);
      toast.error(`Une erreur est survenue lors de la mise à jour de votre compte : ${detail}`);
    }
  });

  const form = useForm<OwnAccountUpdate>({
    defaultValues: {
      username: user?.username,
      firstName: user?.firstName,
      lastName: user?.lastName
    }
  });

  const onSubmit = (data: OwnAccountUpdate) => {
    updateOwnAccount.mutate({
      body: data
    });
  };

  return (
    <Form {...form}>
      <form
        className='grow flex justify-between flex-col gap-2'
        onSubmit={form.handleSubmit(onSubmit)}
      >
        <FormField
          control={form.control}
          name='firstName'
          render={({ field }) => (
            <FormItem>
              <FormLabel>Prénom</FormLabel>
              <FormControl>
                <Input
                  placeholder='Prénom'
                  {...field}
                  value={field.value ?? ''}
                />
              </FormControl>
              <FormDescription>Modification du prénom</FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name='lastName'
          render={({ field }) => (
            <FormItem>
              <FormLabel>Nom de famille</FormLabel>
              <FormControl>
                <Input
                  placeholder='Nom de famille'
                  {...field}
                  value={field.value ?? ''}
                />
              </FormControl>
              <FormDescription>Modification du nom de famille</FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name='username'
          render={({ field }) => (
            <FormItem>
              <FormLabel>Nom d&apos;utilisateur</FormLabel>
              <FormControl>
                <Input
                  placeholder="Nom d'utilisateur"
                  {...field}
                  value={field.value ?? ''}
                />
              </FormControl>
              <FormDescription>Modification du nom d&apos;utilisateur</FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <div className='grow'></div>
        <Button
          type='submit'
          loading={updateOwnAccount.isLoading}
        >
          Modifier mon compte
        </Button>
      </form>
    </Form>
  );
}
