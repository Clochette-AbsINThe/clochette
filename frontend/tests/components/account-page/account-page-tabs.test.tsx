import { render } from '@testing-library/react';

import { findActiveLink, filterLinks, Tabs } from '@/components/account-page/account-page-tabs';

describe('findActiveLink', () => {
  it('should return the active link', () => {
    const pathname = '/account/dashboard';
    const activeLink = findActiveLink(pathname);
    expect(activeLink?.href).toBe('/account/dashboard');
  });

  it('should return undefined if no active link is found', () => {
    const pathname = '/account/settings';
    const activeLink = findActiveLink(pathname);
    expect(activeLink).toBeUndefined();
  });
});

describe('filterLinks', () => {
  it('should return all links if session is null', () => {
    const session = null;
    const filteredLinks = filterLinks(session);
    expect(filteredLinks).toEqual([]);
  });

  it('should return only links with matching scopes', () => {
    const session = {
      scopes: ['president']
    };
    const filteredLinks = filterLinks(session as any);
    expect(filteredLinks).toEqual([
      { label: 'Accueil', href: '/account', scopes: ['staff', 'treasurer', 'president'] },
      { label: 'Tableau de bord', href: '/account/dashboard', scopes: ['president', 'treasurer'] },
      { label: 'Comptes utilisateurs', href: '/account/users', scopes: ['president'] },
      { label: 'TrÃ©sorie', href: '/account/tresory', scopes: ['president', 'treasurer'] }
    ]);
  });

  it('should return an empty array if no links match the user roles', () => {
    const session = {
      scopes: ['member']
    };
    const filteredLinks = filterLinks(session as any);
    expect(filteredLinks).toEqual([]);
  });
});

describe('Tabs', () => {
  it('should render the tabs', () => {
    const { container } = render(<Tabs />);
    expect(container).toMatchSnapshot();
  });
});
