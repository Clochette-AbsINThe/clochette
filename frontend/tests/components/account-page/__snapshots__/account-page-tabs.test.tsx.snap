// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > should render the tabs 1`] = `
<div>
  <div
    class="flex justify-center pr-2 md:border-r md:border-b-0 border-b md:mr-4 mb-4 md:mb-0"
  >
    <div
      class="w-full max-w-md md:w-max flex justify-start items-center md:items-start md:flex-col flex-row mb-2 md:mb-0"
    >
      <a
        class="px-3 py-2 text-sm w-full font-medium bg-slate-400 dark:bg-gray-800"
        href="/account"
      >
        Accueil
      </a>
      <a
        class="px-3 py-2 text-sm w-full font-medium text-gray-900 hover:bg-slate-300 dark:text-gray-300 dark:hover:bg-gray-700"
        href="/account/dashboard"
      >
        Tableau de bord
      </a>
      <a
        class="px-3 py-2 text-sm w-full font-medium text-gray-900 hover:bg-slate-300 dark:text-gray-300 dark:hover:bg-gray-700"
        href="/account/users"
      >
        Comptes utilisateurs
      </a>
      <a
        class="px-3 py-2 text-sm w-full font-medium text-gray-900 hover:bg-slate-300 dark:text-gray-300 dark:hover:bg-gray-700"
        href="/account/tresory"
      >
        Tr√©sorie
      </a>
    </div>
  </div>
</div>
`;
